Methylome.At_main <- function(var1, # control
                              var2, # treatment
                              var1_path,
                              var2_path,
                              Methylome.At_path = ".",
                              annotation_file = "./annotation_files/Methylome.At_annotations.csv.gz",
                              description_file = "./annotation_files/Methylome.At_annotations.csv.gz",
                              TEs_file = "./annotation_files/Methylome.At_annotations.csv.gz",
                              minProportionDiff = c(0.4, 0.2, 0.1), # CG, CHG, CHH
                              binSize = 100,
                              minCytosinesCount = 4,
                              minReadsPerCytosine = 4,
                              pValueThreshold = 0.05,
                              methyl_files_type = "CX_report",
                              img_type = "pdf",
                              n.cores = 8,
                              run_PCA_plot = TRUE,
                              run_total_meth_plot = TRUE,
                              run_CX_Chrplot = TRUE,
                              run_TF_motifs = TRUE,
                              run_GO_analysis = FALSE,
                              run_KEGG_pathways = FALSE,
                              analyze_dH = FALSE,
                              run_TE_metaPlots = FALSE,
                              run_GeneBody_metaPlots = FALSE,
                              run_GeneFeatures_metaPlots = FALSE,
                              gene_features_binSize = 10,
                              metaPlot.random.genes = 10000) {

  ###########################################################################

  start_time <- Sys.time()
  time_msg <<- function(suffix = "\t") paste0(format(Sys.time(), "[%H:%M]"), suffix)
  sep_cat <- function(x, short = F) paste0("\n---- ", x, " ", paste(rep("-", ifelse(short, 20, 50) - nchar(x)), collapse = ""), "\n")
  scripts_dir <- paste0(Methylome.At_path, "/scripts")

  # source all R scripts
  script_files <- list.files(scripts_dir, pattern = "\\.R$", full.names = TRUE)
  script_files <- script_files[!grepl("install_R_packages\\.R$", script_files)]
  script_files <- script_files[!grepl("Methylome\\.At_run\\.R$", script_files)]
  script_files <- script_files[!grepl("Methylome\\.At_main\\.R$", script_files)]
  script_files <- script_files[!grepl("MetaPlots_run\\.R$", script_files)]
  script_files <- script_files[!grepl("mean_deltaH_CX\\.R", script_files)] # have match functions with 'ChrPlots' functions
  script_files <- script_files[!grepl("ChrPlots_", script_files)]
  if (!analyze_dH) {
    script_files <- script_files[!grepl("delta_H_option\\.R$", script_files)]
  }

  invisible(lapply(script_files, source))

  ###########################################################################

  formals(img_device)$img_type <- img_type

  ###########################################################################

  setwd(Methylome.At_path)

  ##### read annotation and description files #####
  cat("\rload annotations and description files [0/3]")

  # annotation file
  tryCatch(
    {
      # if its 'csv' file
      if (grepl("\\.csv$|\\.csv\\.gz$", annotation_file)) {
        annotation.gr <- read.csv(annotation_file) %>%
          makeGRangesFromDataFrame(., keep.extra.columns = T) %>%
          trimm_and_rename()
        # if its 'gtf'/'gff'/'gff3' file
      } else if (grepl("\\.gtf$|\\.gff$|\\.gff3$|\\.gtf\\.gz$|\\.gff\\.gz$|\\.gff3\\.gz$", tolower(annotation_file))) {
        annotation.gr <- import.gff3(annotation_file) %>%
          trimm_and_rename()
      }
      message(time_msg(), "load annotation file")
    },
    error = function(cond) {
      cat("\n*\n load annotation file:\n", as.character(cond), "*\n")
      message(time_msg(), "load 'annotation' file: fail")
    }
  )
  cat("\rload annotations and description files [1/3]")

  # TAIR10 Transposable Elements file
  tryCatch(
    {
      TE_file.df <- read.csv(TEs_file, sep = "\t")
      TE_file <- edit_TE_file(TE_file.df)
      message(time_msg(), "load Transposable Elements file")
    },
    error = function(cond) {
      cat("\n*\n load TE file:\n", as.character(cond), "*\n")
      message(time_msg(), "load Transposable Elements file: fail")
    }
  )
  cat("\rload annotations and description files [2/3]")

  # upload description file
  tryCatch(
    {
      des_file_sep <- ifelse(grepl("\\.csv$|\\.csv\\.gz$", description_file), ",", "\t")
      description_df <- read.csv(description_file, sep = des_file_sep)
      names(description_df)[1] <- "gene_id"
      message(time_msg(), "load description file\n")
    },
    error = function(cond) {
      cat("\n*\n load description file:\n", as.character(cond), "*\n")
      message(time_msg(), "load description file: fail\n")
    }
  )
  cat("\rload annotations and description files [3/3]")

  cat("\n\n")

  ###########################################################################
  setwd(Methylome.At_path)

  is_single <- (length(var1_path) == 1 & length(var2_path) == 1) # both genotypes includes 1 sample
  is_Replicates <- (length(var1_path) > 1 & length(var2_path) > 1) # both genotypes includes >1 samples

  var_args <- list(
    list(path = var1_path, name = var1),
    list(path = var2_path, name = var2)
  )

  ##### load methylation data ('CX_report' file) #####
  message(time_msg(), "load CX methylation data")
  tryCatch(
    {
      # load 'CX_reports'
      n.cores.load <- ifelse(n.cores > 1, 2, 1)
      load_vars <- mclapply(var_args, function(x) {
        Sys.sleep(ifelse(x$name == var1, 0, 2))
        load_replicates(x$path, n.cores, x$name, F, methyl_files_type)
      },
      mc.cores = n.cores.load
      )

      # trimm seqs objects (rename if not 'TAIR10' Chr seqnames)
      meth_var1 <- trimm_and_rename(load_vars[[1]]$methylationData_pool)
      meth_var2 <- trimm_and_rename(load_vars[[2]]$methylationData_pool)
      meth_var1_replicates <- trimm_and_rename(load_vars[[1]]$methylationDataReplicates)
      meth_var2_replicates <- trimm_and_rename(load_vars[[2]]$methylationDataReplicates)
      cat("\nPooled ", var1, " object for example:\n\n", sep = "")
      capture.output(print(meth_var1), file = NULL)[-c(1, 16)] %>% cat(sep = "\n")
      cat(paste(" seq-levels:", paste(seqlevels(meth_var1), collapse = " ")), "\n\n")

      if (!is_single) {
        # join replicates
        methylationDataReplicates_joints <- joinReplicates(
          meth_var1_replicates,
          meth_var2_replicates
        )
        message(time_msg(), "load and join replicates data: successfully")
      } else {
        # join singles
        methylationDataReplicates_joints <- joinReplicates(
          meth_var1,
          meth_var2
        )
        message(time_msg(), "load and join single-samples data: successfully")
      }
    },
    error = function(cond) {
      cat("\n*\n load and join CX methylation data:\n", as.character(cond), "*\n")
      stop("load and join CX methylation data: fail")
    }
  )

  ###########################################################################

  # new folders path names
  comparison_name <- paste0(var2, "_vs_", var1)
  exp_path <- paste0(Methylome.At_path, "/results/", comparison_name)
  PCA_plots_path <- paste0(exp_path, "/PCA_plots")
  meth_levels_path <- paste0(exp_path, "/methylation_levels")
  ChrPlot_CX_path <- paste0(exp_path, "/ChrPlot_CX")
  ChrPlot_subCX_path <- paste0(exp_path, "/ChrPlot_CX/subCX")
  TF_motifs_path <- paste0(exp_path, "/TF_motifs")
  gainORloss_path <- paste0(exp_path, "/gain_OR_loss")
  genome_ann_path <- paste0(exp_path, "/genome_annotation")
  DMRs_bigWig_path <- paste0(exp_path, "/DMRs_bigWig")
  ChrPlots_DMRs_path <- paste0(exp_path, "/ChrPlot_DMRs")
  dH_CX_path <- paste0(exp_path, "/deltaH")
  dH_CX_ann_path <- paste0(exp_path, "/deltaH/genome_annotation")
  DMV_analysis_path <- paste0(exp_path, "/DMV_analysis")
  metaPlot_path <- paste0(exp_path, "/metaPlots")

  TAIR10_TFBS_file <- paste0(Methylome.At_path, "/annotation_files/TAIR10_compressed_TFBSs.bed.gz")

  dir.create(exp_path, showWarnings = F)
  setwd(exp_path)

  ###########################################################################

  message(sep_cat("QC"))
  cat(sep_cat("QC"))

  ##### calculate the conversion rate by the chloroplast chromosome (ChrC)
  message(time_msg(), "conversion rate (C->T) along the Chloroplast genome:", appendLF = F)
  cat("\nconversion rate (C->T) along the Chloroplast genome:") # "\n"
  tryCatch(
    {
      message("")
      conR_var1 <- conversionRate(load_vars[[1]]$methylationDataReplicates, var1)
      conR_var2 <- conversionRate(load_vars[[2]]$methylationDataReplicates, var2)
      conR_b <- rbind(conR_var1, conR_var2)
      write.csv(conR_b, paste0(exp_path, "/conversion_rate.csv"), row.names = F)
      print(kable(conR_b))
    },
    error = function(cond) {
      cat("\n*\n conversion rate:\n", as.character(cond), "*\n")
      message("fail")
      cat(" fail\n")
    }
  )

  rm(load_vars)
  message(sep_cat("Total methylation"))
  cat(sep_cat("Total methylation"))

  ##### PCA plot to total methlyation in all contexts
  if (run_PCA_plot) {
    if (!is_single) {
      dir.create(PCA_plots_path, showWarnings = F)
      setwd(PCA_plots_path)

      message(time_msg(), "generating PCA plots of total methylation levels: ", appendLF = F)
      cat("\nPCA plots...")
      tryCatch(
        {
          pca_plot(methylationDataReplicates_joints, var1, var2, var1_path, var2_path, "CG")
          pca_plot(methylationDataReplicates_joints, var1, var2, var1_path, var2_path, "CHG")
          pca_plot(methylationDataReplicates_joints, var1, var2, var1_path, var2_path, "CHH")
          pca_plot(methylationDataReplicates_joints, var1, var2, var1_path, var2_path, "all_contexts")
          message("done")
        },
        error = function(cond) {
          cat("\n*\n PCA plot:\n", as.character(cond), "*\n")
          message("fail")
        }
      )
      setwd(exp_path)
      cat(" done\n")
    } else {
      message(time_msg(), "* skipping PCA plots for single-samples data")
      cat("skipping PCA plots for single-samples data\n")
    }
  }

  ##### calculate and plot total methylation levels (%)
  if (run_total_meth_plot) {
    dir.create(meth_levels_path, showWarnings = F)
    setwd(meth_levels_path)

    message(time_msg(), "bar-plots for total methylation levels (5-mC%): ", appendLF = F)
    cat("bar-plots for total methylation levels...")
    tryCatch(
      {
        total_meth_levels <- total_meth_levels(meth_var1_replicates, meth_var2_replicates, var1, var2)
        message("done")
        cat(" done\n")
      },
      error = function(cond) {
        cat("\n*\n total methylation levels:\n", as.character(cond), "*\n")
        message("fail")
        cat(" fail\n")
      }
    )
  }

  ###########################################################################

  ##### ChrPlots for CX methylation #####
  if (run_CX_Chrplot) {
    dir.create(ChrPlot_CX_path, showWarnings = F)
    dir.create(ChrPlot_subCX_path, showWarnings = F)
    setwd(ChrPlot_CX_path)

    cat("\n* chromosome methylation plots (ChrPlots):")
    message(time_msg(), "generating chromosome methylation plots (ChrPlots): ", appendLF = F)
    tryCatch(
      {
        source(paste0(scripts_dir, "/ChrPlots_CX.R"))
        suppressWarnings(run_ChrPlots_CX(var1, var2, meth_var1, meth_var2, TE_file, n.cores))
        message("done")
      },
      error = function(cond) {
        cat("\n*\n ChrPlots:\n", as.character(cond), "*\n")
        message("fail")
      }
    )
  }

  setwd(exp_path)

  ###########################################################################

  ##### Transcription facrots motif analysis (UniBind - TFBS) #####
  if (run_TF_motifs) {
    dir.create(TF_motifs_path, showWarnings = F)
    setwd(TF_motifs_path)

    cat("\nTranscription factors motifs plot:\n")
    message(time_msg(), "generating transcription factors motifs plots")
    tryCatch(
      {
        suppressWarnings(TF_motifs(methylationDataReplicates_joints, "all", 1e6, NULL, annotation.gr, TAIR10_TFBS_file))
        # message("done")
      },
      error = function(cond) {
        cat("\n*\n nTranscription factors motifs plot:\n", as.character(cond), "*\n")
        message("fail")
      }
    )
  }

  setwd(exp_path)

  ###########################################################################


  ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###
  ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###
  ### ### DMRs and its downstream results ### ### ### ### ### ###

  message(sep_cat("DMRs analysis"))
  cat(sep_cat("DMRs analysis"))

  ##### call DMRs for replicates/single data
  message(time_msg(), "call DMRs for replicates data: ", is_Replicates)
  cat(paste0("\n", time_msg(" "), "call DMRs for replicates data: ", is_Replicates, "\n"))

  if (is_Replicates) {
    cat(paste0(time_msg(" "), "compute ", binSize, "bp DMRs using: beta regression\n"))
    message(time_msg(), "compute ", binSize, "bp DMRs using: beta regression\n")
  } else {
    cat(paste0(time_msg(" "), "compute ", binSize, "bp DMRs using: fisher’s exact test\n"))
    message(time_msg(), "compute ", binSize, "bp DMRs using: fisher’s exact test\n")
  }

  ##############################
  ##### Calling DMRs in Replicates #####
  DMRs_results <- mclapply(c("CG", "CHG", "CHH"), function(context) {
    setwd(exp_path)
    tryCatch(
      {
        DMRs_call <- calling_DMRs(
          methylationDataReplicates_joints, meth_var1, meth_var2,
          var1, var2, var1_path, var2_path, comparison_name,
          context, minProportionDiff, binSize, pValueThreshold,
          minCytosinesCount, minReadsPerCytosine, ifelse(n.cores > 3, n.cores / 3, 1), is_Replicates
        )

        # quatiles cutoff for dH analysis
        if (analyze_dH) {
          DMRs_call <- proportions_cutoff(DMRs_call, meth_var1_replicates, context, q = 0.99)
        }

        cat(paste0(time_msg(" "), "statistically significant DMRs (", context, "): ", length(DMRs_call), "\n"))
        message(time_msg(), paste0("statistically significant DMRs (", context, "): ", length(DMRs_call)))
        # message(time_msg(), paste0("\tDMRs caller in ", context, " context: done"))
        return(DMRs_call)
      },
      error = function(cond) {
        cat(paste0("\n*\n Calling DMRs in ", context, " context:\n"), as.character(cond), "*\ncontinue without calling DMRs!\n\n")
        message(time_msg(), "\tCalling DMRs: fail\n")
        return(NULL)
      }
    )
  }, mc.cores = ifelse(n.cores >= 3, 3, 1))

  names(DMRs_results) <- c("CG", "CHG", "CHH")
  cat(paste0(time_msg(" "), "done!\n"))
  message("")

  ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###
  ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###
  ### ### main loop for plots ### ### ### ### ### ### ### ### ###

  DMRs_ann_plots_list <- list(CG = NULL, CHG = NULL, CHH = NULL)
  te_vs_gene_bar <- list(CG = NULL, CHG = NULL, CHH = NULL)

  for (context in c("CG", "CHG", "CHH")) {
    cat(sep_cat(paste(context, "annotations"), T), "\n")

    DMRs_bins <- DMRs_results[[context]]

    # skip if DMRs calling failed
    if (is.null(DMRs_bins)) {
      message(time_msg(paste0("\t", context, ":")), paste0("\tSkipping ", context, " - no DMRs available"))
      next
    }

    ##############################
    #####  Gain or Loss - DMRs #####
    dir.create(gainORloss_path, showWarnings = FALSE)
    setwd(gainORloss_path)

    ##### Pie chart
    tryCatch(
      {
        gainORloss(DMRs_bins, context, add_count = T)
        message(time_msg(paste0("\t", context, ":")), "\tpie chart (gain or loss): done")
      },
      error = function(cond) {
        cat("\n*\n pie chart (gain or loss):\n", as.character(cond), "*\n")
        message(time_msg(paste0("\t", context, ":")), "\tpie chart (gain or loss): fail\n")
      }
    )

    ##### Ratio distribution
    tryCatch(
      {
        ratio.distribution(DMRs_bins, var1, var2, context, comparison_name)
        message(time_msg(paste0("\t", context, ":")), "\tratio distribution (gain or loss): done")
      },
      error = function(cond) {
        cat("\n*\n ratio distribution (gain or loss):\n", as.character(cond), "*\n")
        message(time_msg(paste0("\t", context, ":")), "\tratio distribution (gain or loss): fail\n")
      }
    )
    ##############################

    ##### ChrPlots for DMRs #####
    tryCatch(
      {
        source(paste0(scripts_dir, "/ChrPlots_DMRs.R"))
        dir.create(ChrPlots_DMRs_path, showWarnings = FALSE)
        setwd(ChrPlots_DMRs_path)
        ChrPlots_DMRs(comparison_name, DMRs_bins, var1, var2, context, scripts_dir)
        message(time_msg(paste0("\t", context, ":")), "\tgenerated ChrPlots for all DMRs: done")
      },
      error = function(cond) {
        cat("\n*\n generated ChrPlots for all DMRs:\n", as.character(cond), "*\n")
        message(time_msg(paste0("\t", context, ":")), "\tgenerated ChrPlots for all DMRs: fail\n")
      }
    )
    setwd(exp_path)

    ##### Annotate DMRs and total-methylations #####
    cat("genome annotations for", context, "DMRs:\n")
    message(time_msg(paste0("\t", context, ":")), "\tgenome annotations for DMRs...")
    dir.create(genome_ann_path, showWarnings = FALSE)
    setwd(genome_ann_path)

    # genome annotations
    tryCatch(
      {
        ann_list <- genome_ann(annotation.gr, TE_file) # create annotations from annotation file as a list
        DMRs_ann(ann_list, DMRs_bins, context, description_df) # save tables of annotate DMRs. have to run after 'genome_ann'
        CX_ann(ann_list, var1, var2, meth_var1, meth_var2, context) # save tables of annotate CX
        DMRs_ann_plots_list[[context]] <- DMRs_ann_plots(var1, var2, context)
        message(time_msg(paste0("\t", context, ":")), "\tgenome annotations for DMRs: done")
      },
      error = function(cond) {
        cat("\n*\n genome annotations for DMRs:\n", as.character(cond), "*\n")
        message(time_msg(paste0("\t", context, ":")), "\tgenome annotations for DMRs: fail")
      }
    )

    # additional TE annotations results
    tryCatch(
      {
        te_vs_gene_bar[[context]] <- TE_ann_plots(context, TE_file)
        TE_Super_Family_Frequency(context, TE_file)
      },
      error = function(cond) {
        cat("\n*\n TE families plots:\n", as.character(cond), "*\n")
        message(time_msg(paste0("\t", context, ":")), "\tTE families plots: fail")
      }
    )
    setwd(exp_path)

    ##### save DMRs as bigWig file #####
    dir.create(DMRs_bigWig_path, showWarnings = FALSE)
    setwd(DMRs_bigWig_path)
    for (cntx_g2b in c("CG", "CHG", "CHH")) {
      suppressWarnings(try(
        {
          gr_2_bigWig(DMRs_results[[cntx_g2b]], paste0(paste("DMRs", cntx_g2b, comparison_name, sep = "_"), ".bw"))
        },
        silent = T
      ))
    }
    message(time_msg(paste0("\t", context, ":")), "\tsaved all DMRs also as bigWig files\n")
    cat("saved all DMRs also as bigWig files\n")
  }

  setwd(exp_path)

  ### ### # finish main loop  ### ### ### ### ### ### ### ### ###
  ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###
  ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###

  ###########################################################################

  message(sep_cat("Downstream DMRs"))
  cat(sep_cat("Downstream DMRs"))

  # save gene-body DMRs bar-plot
  tryCatch(
    {
      combined_plot <- plot_grid(
        DMRs_ann_plots_list$CG,
        DMRs_ann_plots_list$CHG,
        DMRs_ann_plots_list$CHH,
        gene_ann_legend(),
        nrow = 1,
        rel_widths = c(2.45, 2.45, 2.45, 1.34)
      )

      img_device(paste0("genome_annotation/DMRs_genome_annotation_", comparison_name), w = 8.69, h = 2)
      print(combined_plot)
      dev.off()
    },
    error = function(cond) {
      cat("\n*\n save gene-body DMRs bar-plot: fail\n", as.character(cond), "*\n")
      message(time_msg(), "save gene-body DMRs bar-plot: fail")
    }
  )

  # save TE vs gene bar-plot
  tryCatch(
    {
      combined_te_plot <- plot_grid(
        te_vs_gene_bar$CG,
        te_vs_gene_bar$CHG,
        te_vs_gene_bar$CHH,
        nrow = 1
      )

      img_device(paste0("genome_annotation/DMRs_TEs_vs_coding_genes_", comparison_name), w = 4.8, h = 2.1)
      print(combined_te_plot)
      dev.off()
    },
    error = function(cond) {
      cat("\n*\n save TE vs gene bar-plot: fail\n", as.character(cond), "*\n")
      message(time_msg(), "save TE vs gene bar-plot: fail")
    }
  )

  ###########################################################################

  ##### DMRs density - curcular plot #####
  tryCatch(
    {
      setwd(exp_path)
      cat("\ngenerated DMRs density plot for all contexts: ")
      # setwd(ChrPlots_DMRs_path)
      DMRs_circular_plot(annotation.gr, TE_file, comparison_name)
      cat("done\n")
      message(time_msg(), "generated DMRs density plot for all contexts: done")
    },
    error = function(cond) {
      cat("\n*\n DMRs density plot:\n", as.character(cond), "*\n")
      message(time_msg(), "generated DMRs density plot for all contexts: fail")
    }
  )

  ###########################################################################

  ##### TEs superfamily - curcular plot #####
  tryCatch(
    {
      setwd(genome_ann_path)
      cat("generated DMRs density plots over TEs superfamilies: ")
      TEs_superfamily_circular_plot(annotation.gr)
      cat("done\n")
      message(time_msg(), "generated DMRs over TEs superfamilies: done")
    },
    error = function(cond) {
      cat("\n*\n DMRs over TEs superfamilies:\n", as.character(cond), "*\n")
      message(time_msg(), "generated DMRs over TEs superfamilies: fail")
    }
  )

  setwd(exp_path)

  ###########################################################################

  ##### DMRs within Genes-groups #####
  func_groups_path <- paste0(genome_ann_path, "/functional_groups")
  setwd(exp_path)
  dir.create(func_groups_path, showWarnings = FALSE)
  cat("Annotate DMRs into functional groups: ")
  for (ann.l in c("Genes", "Promoters")) {
    groups_results <- mclapply(c("CG", "CHG", "CHH", "all"), function(cntx.l) {
      tryCatch(
        {
          DMRs_into_groups(treatment = comparison_name, ann = ann.l, context = cntx.l)
        },
        error = function(cond) {
          cat("\n*\n Annotate *", cntx.l, "* - *", ann.l, "* DMRs into functional groups:\n", as.character(cond), "*\n")
          message(paste("Annotate", cntx.l, "-", ann.l, "DMRs into functional groups: fail\n"))
          return(NULL)
        }
      )
    }, mc.cores = ifelse(n.cores >= 4, 4, n.cores)) %>%
      do.call(rbind, .) %>%
      arrange(context)

    tryCatch(
      {
        img_device(paste0(func_groups_path, "/", ann.l, "_groups_barPlots_", comparison_name), w = 14, h = 4.5)
        print(groups_barPlots(groups_results))
        dev.off()
      },
      error = function(cond) {
        cat("\n*\n Bar-plot of *", cntx.l, "* - *", ann.l, "* annotated DMRs into functional groups:\n", as.character(cond), "*\n")
      }
    )
  }
  cat("done\n")
  message(time_msg(), "Annotate DMRs into functional groups: done")

  ###########################################################################

  ##### TEs - size and distance plots
  tryCatch(
    {
      TE_context_list <- TE_delta_meth(list(meth_var1, meth_var2), TE_file)

      dir.create(paste0(genome_ann_path, "/TEs_addiotionnal_results/TE_size_n_distance/"), showWarnings = FALSE)

      ## TE methylation levels (delta) and size
      cat("\nTE delta-methylation vs. TE-size\n")
      for (te_sz_cntx in c("CG", "CHG", "CHH")) {
        ggsave(
          filename = paste0(genome_ann_path, "/TEs_addiotionnal_results/TE_size_n_distance/", te_sz_cntx, "_TE_size_delta_scatter.png"),
          plot = te_size_plot(TE_context_list, te_sz_cntx),
          width = 3750,
          height = 2500,
          units = "px",
          dpi = 1200
        )
      }
      message(time_msg(), "TE delta-methylation vs. TE-size: done")

      ## TE methylation levels (delta) and distance from centromer
      cat("TE delta-methylation vs. distance from centromere\n")
      TE_distance <- distance_from_centromer(TE_context_list, TE_file, window_size = 1e6)

      ggsave(
        filename = paste0(genome_ann_path, "/TEs_addiotionnal_results/TE_size_n_distance/TE_centromere_distance_delta.png"),
        plot = TE_distance$plot,
        width = 3750,
        height = 2500,
        units = "px",
        dpi = 1200
      )
      message(time_msg(), "TE delta-methylation vs. distance from centromer: done\n")
    },
    error = function(cond) {
      cat("\n*\n TEs size and distance plots:\n", as.character(cond), "*\n")
    }
  )

  ###########################################################################


  ##### GO analysis for annotated DMRs
  if (run_GO_analysis) {
    cat(sep_cat("GO analysis"))
    tryCatch(
      {
        GO_path <- paste0(exp_path, "/GO_analysis")
        dir.create(GO_path, showWarnings = FALSE)

        message(time_msg(), "GO analysis for annotated DMRs...")
        run_GO(comparison_name, genome_ann_path, GO_path, n.cores)
        message(time_msg(), "GO analysis for annotated DMRs: done\n")
      },
      error = function(cond) {
        cat("\n*\n GO analysis:\n", as.character(cond), "*\n")
        message(time_msg(), "GO analysis for annotated DMRs: fail\n")
      }
    )
  }

  ##### KEGG pathways for annotated DMRs
  if (run_KEGG_pathways) {
    cat(sep_cat("KEGG pathways"))
    tryCatch(
      {
        KEGG_path <- paste0(exp_path, "/KEGG_pathway")
        dir.create(KEGG_path, showWarnings = FALSE)

        message(time_msg(), "KEGG pathways for annotated DMRs...")
        run_KEGG(comparison_name, genome_ann_path, KEGG_path, n.cores)
        message(time_msg(), "KEGG pathways for annotated DMRs: done\n")
      },
      error = function(cond) {
        cat("\n*\n KEGG pathways:\n", as.character(cond), "*\n")
        message(time_msg(), "KEGG pathways for annotated DMRs: fail\n")
      }
    )
  }

  ###########################################################################

  ##### dH analysis over CX methylation #####
  if (analyze_dH) {
    message(sep_cat("ΔH analysis"))
    cat(sep_cat("ΔH analysis"))
    dir.create(dH_CX_path, showWarnings = F)
    dir.create(dH_CX_ann_path, showWarnings = F)
    setwd(dH_CX_path)

    cat("\n* chromosome dH plots (ChrPlots):")
    message(time_msg(), "generating ChrPlot (mean of dH) and scatter-plot (dH vs dmC): ", appendLF = F)
    tryCatch(
      {
        source(paste0(scripts_dir, "/mean_deltaH_CX.R"))
        suppressWarnings(run_mean_deltaH_CX(var1, var2, meth_var1, meth_var2, TE_file, n.cores))
        message("done")
      },
      error = function(cond) {
        cat("\n*\n mean dH:\n", as.character(cond), "*\n")
        message("fail")
      }
    )

    # message(time_msg(), "generating sum dH analysis:\n", appendLF = F) # , rep("-", 29)
    # tryCatch(
    #   {
    #     suppressWarnings(run_sum_deltaH_CX(var1, var2, meth_var1, meth_var2, annotation.gr, TE_file, description_df, n.cores, fdr = 0.95))
    #   },
    #   error = function(cond) {
    #     cat("\n*\n sum dH:\n", as.character(cond), "*\n")
    #     message("fail\n")
    #   }
    # )
  }

  setwd(exp_path)

  ###########################################################################

  message(sep_cat("DMV analysis"))
  cat(sep_cat("DMV analysis\n"))

  ##### Calling DMVs in Replicates #####
  dir.create(DMV_analysis_path)
  setwd(DMV_analysis_path)
  DMVs_results <- mclapply(c("CG", "CHG", "CHH"), function(context) {
    tryCatch(
      {
        DMVs_call <- calling_DMRs(
          methylationDataReplicates_joints, meth_var1, meth_var2,
          var1, var2, var1_path, var2_path, comparison_name,
          context, minProportionDiff,
          binSize = 1000, pValueThreshold,
          minCytosinesCount = 20, minReadsPerCytosine = 5, ifelse(n.cores > 3, n.cores / 3, 1),
          is_Replicates, analysis_name = "DMVs"
        )
        cat(paste0(time_msg(" "), "statistically significant DMVs (", context, "): ", length(DMVs_call), "\n"))
        message(time_msg(), paste0("statistically significant DMVs (", context, "): ", length(DMVs_call)))
        # message(time_msg(), paste0("\tDMVs caller in ", context, " context: done"))
        return(DMVs_call)
      },
      error = function(cond) {
        cat(paste0("\n*\n Calling DMVs in ", context, " context:\n"), as.character(cond), "*\ncontinue without calling DMVs!\n\n")
        message(time_msg(), "\tCalling DMVs: fail\n")
        return(NULL)
      }
    )
  }, mc.cores = ifelse(n.cores >= 3, 3, 1))

  names(DMVs_results) <- c("CG", "CHG", "CHH")
  cat(paste0(time_msg(" "), "done!\n"))

  ##### save DMRs as bigWig file #####
  setwd(DMV_analysis_path)
  for (cntx_g2b in c("CG", "CHG", "CHH")) {
    suppressWarnings(try(
      {
        gr_2_bigWig(DMVs_results[[cntx_g2b]], paste0(DMV_analysis_path, paste("/DMVs", cntx_g2b, comparison_name, sep = "_"), ".bw"))
      },
      silent = T
    ))
  }
  message(time_msg(), "saved all DMRs also as bigWig files\n")
  cat("saved all DMVs also as bigWig files\n")

  message("")

  ###########################################################################

  ##### run metPlot function for coding-Genes and TEs
  if (run_TE_metaPlots | run_GeneBody_metaPlots | run_GeneFeatures_metaPlots) {
    message(sep_cat("MetaPlots"))
    cat(sep_cat("MetaPlots"))
    # cat(sep_cat, "\nmetaPlots:\n----------")
    dir.create(metaPlot_path, showWarnings = F)

    # calculate metaPlot for genes bodies
    if (run_TE_metaPlots) {
      tryCatch(
        {
          message(paste("generate metaPlot to", metaPlot.random.genes, "protein-coding Genes..."))
          setwd(metaPlot_path)
          Genes_metaPlot(meth_var1, meth_var2, var1, var2, annotation.gr, metaPlot.random.genes, minReadsPerCytosine, n.cores, is_TE = F)
          setwd(metaPlot_path)
          delta_metaplot("Genes", var1, var2)
        },
        error = function(cond) {
          cat("\n*\n TEs metaPlots:\n", as.character(cond), "*\n")
          message(time_msg(), paste0("process average metaPlot to ", metaPlot.random.genes, " Protein Coding Genes: fail"))
        }
      )
    }

    # calculate metaPlot for TEs
    if (run_GeneBody_metaPlots) {
      tryCatch(
        {
          message(paste("\ngenerate metaPlot to", metaPlot.random.genes, " Transposable Elements..."))
          setwd(metaPlot_path)
          Genes_metaPlot(meth_var1, meth_var2, var1, var2, TE_file, metaPlot.random.genes, minReadsPerCytosine, n.cores, is_TE = T)
          setwd(metaPlot_path)
          delta_metaplot("TEs", var1, var2)
        },
        error = function(cond) {
          cat("\n*\n Transposable Elements metaPlots:\n", as.character(cond), "*\n")
          message(time_msg(), paste0("process average metaPlot to ", metaPlot.random.genes, " Transposable Elements: fail\n"))
        }
      )
    }

    # calculate metaPlot for coding-Gene features (CDS, introns, etc.)
    if (run_GeneFeatures_metaPlots) {
      tryCatch(
        {
          message(paste("\ngenerate metaPlot to", metaPlot.random.genes, "protein-coding Gene Features..."))
          setwd(metaPlot_path)
          Genes_features_metaPlot(meth_var1, meth_var2, var1, var2, annotation.gr, metaPlot.random.genes, minReadsPerCytosine, gene_features_binSize, n.cores)
          # delta_metaplot("Gene_features", var1, var2, is_geneFeature = TRUE)
        },
        error = function(cond) {
          cat("\n*\n Gene features metaPlots:\n", as.character(cond), "*\n")
          message(time_msg(), paste0("process average metaPlot to ", metaPlot.random.genes, " Protein Coding Gene Features: fail"))
        }
      )
    }
  }

  setwd(exp_path)

  ###########################################################################

  setwd(Methylome.At_path)
  message(paste0("**\t", var2, " vs ", var1, ": done\n"))
  cat("\n", rep("-", 56))

  ###########################################################################

  # date and time of the end
  message(paste0(
    "**\t",
    paste(format(Sys.time(), "%d"), format(Sys.time(), "%m"), format(Sys.time(), "%Y"), sep = "-"),
    " ", format(Sys.time(), "%H:%M")
  ))
  end_time <- Sys.time()
  time_diff <- difftime(end_time, start_time, units = "mins") %>% as.numeric()
  end_msg <- paste(
    "**\ttime:",
    paste0(floor(time_diff / 60), "hr"),
    paste0(floor(time_diff %% 60), "min\n")
  )
  cat(paste0("\n\n", time_msg(), "Done!\n", end_msg, "----\n"))
  message(end_msg)
}
